### PR TITLE
Fix #5566: Handle action_required status in code coverage

### LINKED ISSUE #5566: [BUG]: Code Coverage comment fails with disallowed conclusion: action_required
### Describe the bug

The PR #5560, #5559 fails to comment the generated code coverage report.

The [stack trace](https://github.com/oppia/oppia-android/pull/5560/checks?check_run_id=32260222424) mentions:

```
Error: Run#11587680371 is completed with disallowed conclusion: action_required
```

#### Suspect:
Since a review was requested and the trigger relies on the pull request status, it may have marked the action as "required" rather than as a success or failure, which are currently the only two permitted conclusions for the coverage report. (Earlier reviews were submitted as comments, so they didn't lead to this issue.)

**Updated Findings:**

I initially thought the `action_required` status might have been triggered by a review request, but during testing, I was able to replicate it by setting rules to manually require workflow approvals. This makes me think the issue could indeed be related to the approval requirements (notably, both PRs where this occurred required workflow approval). 

Given this, simply adding an `action_required` conclusion or adjusting the check run handling might not fully resolve the issue. None of the reproductions were conclusive, so a more thorough investigation is needed to identify and resolve the underlying cause.

### Steps To Reproduce

- Create a PR and request a review
- Ensure this review is of the type that requires an action ("Request Changes") instead of just adding comments.
- Once the review is submitted, it should be marked as "action_required" in the comment coverage stack trace 

### Expected Behavior

It should get a success conclusion and post a coverage comment with the generated coverage data.

### Screenshots/Videos

_No response_

### What device/emulator are you using?

_No response_

### Which Android version is your device/emulator running?

_No response_

### Which version of the Oppia Android app are you using?

_No response_

### Additional Context

Adding "action_required" to the desired conclusions could be a solution if that's the only cause of this issue, yet it need to be validated.

### PR DESCRIPTION
**Pull request by manas-yu**
## Explanation

Fixes #5566: This PR addresses the issue of code coverage comments not being posted on pull requests when certain review status ("Request Changes") result in an `action_required` status. 

### Changes:
- Updated `comment_coverage.yml` to include `action_required` in the `allowed-conclusions` list for the `check_code_coverage_completed` job. This ensures that the workflow continues even if a review requires further action.
## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

### CODE REVIEW

**Comment by manas-yu** [comment]
Hey @Rd4dev PTAL, the Check code coverage completed job was failing with just adding the action_required in allowed-conclusions so i have added explicit handling of action_required status in check_coverage_results.

**Comment by Rd4dev** [comment]
> Hey @Rd4dev PTAL, the Check code coverage completed job was failing with just adding the action_required in allowed-conclusions so i have added explicit handling of action_required status in check_coverage_results.

@manas-yu, could you link a reproduction of the issue and a demonstration showing the fix in action?

**Comment by manas-yu** [comment]
@Rd4dev could you add a review that requires action ("Request Changes") to reproduce the issue here.

**Comment by Rd4dev** [comment]
@manas-yu, the addition of `action_required` wouldn't directly reflect via this PR as the comment workflow uses a [pull_request_target](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#pull_request_target) trigger, which only runs in the context of the base branch rather than the PR's merge commit. 

As you previously mentioned you tested it
> Check code coverage completed job was failing with just adding the action_required in allowed-conclusions

and it wasn’t working with just the action_required change, I’d suggest getting the added allowed conclusion merged into your base branch where you tested (either fork or separate repo). Then, try replicating the behavior and fix to confirm that it works as expected.

**Comment by Rd4dev** [comment]
@manas-yu Please also take a look at the [latest findings](https://github.com/oppia/oppia-android/issues/5566#issuecomment-2466234835) to see if that could help with reproducing the issue, and if possible, work on a solution based on those insights.

**Comment by manas-yu** [comment]
Hey @Rd4dev, I tried reproducing the issue again using the [test pull request](https://github.com/naoix11/oppia-android/pull/6), and the checks passed even with the "Request Changes" action on the reviews. I’ve reviewed your findings and will provide an update if I discover anything further.

**Comment by Rd4dev** [comment]
@manas-yu, got it! appreciate the update.

**Comment by manas-yu** [comment]
Hey @Rd4dev, I reproduced the issue by enabling workflow approvals. To streamline, we could remove the pull_request_target trigger and use only the workflow_run trigger, which would automatically proceed once the code coverage workflow completes. This approach could simplify the process by triggering actions only when code_coverage.yml is done—whether it needs approval or not. Let me know what you think!

**Comment by Rd4dev** [comment]
@manas-yu, thanks for digging into that! As you found, using `workflow_run` could streamline the workflows and avoid the approval issues. However, the reason we avoided `workflow_run` was because it would lose visibility in the PR checklist. This means it wouldn't show up as part of the PR checks, but would only appear in the Actions tab, which could cause confusion about where to check if the workflow fails.

So far, I’ve tested it, and a manual re-run does make it work and post the comment properly after each workflow approval. But we’ll need a way to automate this if possible.

**Comment by manas-yu** [comment]
Hey @Rd4dev, could you PTAL at  [comment_coverage_report.yml](https://github.com/nifux12/oppia-android/blob/develop/.github/workflows/comment_coverage_report.yml) and [code_coverage_report.yml](https://github.com/nifux12/oppia-android/blob/develop/.github/workflows/code_coverage_report.yml), as well as the 
[test PR](https://github.com/nifux12/oppia-android/pull/5)

I’ve made the following updates:

- Replaced `pull_request_target` with `pull_request` trigger to ensure that Code Coverage Report Analysis runs immediately after `code_coverage.yml`, irrespective of approvals, while maintaining visibility in the PR checklist.

- Created a separate workflow for commenting on the code coverage report, as `peter-evans/create-or-update-comment@v4` does not natively support pull_request triggers.

Let me know if you have any feedback or suggestions. Thanks!

**Comment by manas-yu** [comment]
Should i continue with the above mentioned approach or follow [comment](https://github.com/oppia/oppia-android/issues/5566#issuecomment-2471152767).

**Comment by Rd4dev** [comment]
@manas-yu thanks for taking the time to test this!

Here are a few points to clarify:

1. The `peter-evans/create-or-update-comment@v4` external action would work with `pull_request` triggers; however, in the codebase, it’s set to pull_request_target to allow permissions for posting comments. 

2. Your updated approach is nice but it still uses `workflow_run` to comment the coverage report, which will prevent the workflow from appearing in the PR checklist.

**Suggested Solution:**

This was discussed in today’s CLaM (Oppia Android team) meeting, and the [suggested short-term solution](https://github.com/oppia/oppia-android/issues/5566#issuecomment-2471152767) is:

- To add a clear error message for cases where an `action_required` conclusion occurs, instructing that jobs need to be manually re-run to post the code coverage report after manual approval.

Could you try implementing this workaround and provide a demonstration of it in action.

**Comment by manas-yu** [comment]
I ran into permission issues with `peter-evans/create-or-update-comment@v4` when trying to post comments with `pull_request`, so I ended up using `workflow_run`as a workaround. I’ll go ahead and implement the [suggested solution](https://github.com/oppia/oppia-android/issues/5566#issuecomment-2471152767).👍

**Comment by Rd4dev** [comment]
Ah, that makes sense since it's being tested from a non-upstream branch. Thanks for the update on working through the suggested approach!

**Comment by manas-yu** [comment]
Hey @Rd4dev,
I’ve added a Conclusion Analysis step in the [comment_coverage_report.yml](https://github.com/nifux12/oppia-android/blob/develop/.github/workflows/comment_coverage_report.yml) workflow. This step logs a warning when the code coverage conclusion is action_required, prompting the need for manual approval.

You can see the changes in the [test PR](https://github.com/nifux12/oppia-android/pull/7)

File: .github/workflows/comment_coverage_report.yml
Comment (Rd4dev) [inline .github/workflows/comment_coverage_report.yml]: Please update the warning to an error to fail the step explicitly instead of just issuing a warning. This way, it's clear that we need to re-run the workflow.

Comment (Rd4dev) [inline .github/workflows/comment_coverage_report.yml]: We can update the echo statement slightly to make it clearer, for example: suggestion - "First-time contributor workflows require manual approval. After approval, please re-run the comment coverage workflows to post the coverage report." This clarifies that it's not just the code coverage workflow requiring manual approval and specifies which workflows need to be re-run.

File: .github/workflows/comment_coverage_report.yml
Comment (Rd4dev) [inline .github/workflows/comment_coverage_report.yml]: I believe we can remove the conditional check that skips the step if `action_required` is returned, causing the steps to fail without handling or skipping them, for the same reasons mentioned above. This way, it will be explicitly clear that the steps have failed and will require a re-run, rather than silently passing them.

**Review by Rd4dev — COMMENTED** [review]
Thanks @manas-yu, this was nice, and the demonstration was super helpful! I just had a few suggestions, all of which focus on explicitly failing the steps rather than skipping or passing them. 

The goal is to ensure we can clearly identify which workflows, jobs, and steps have failed so we can easily revisit them after approval and re-run the failed jobs in one go. If we choose to pass or skip them, it could create confusion and give the impression that everything went well when it actually didn’t. 

Once these changes are made and demonstrated, I think this should be good to go! Feel free to drop any suggestions if you have.

**Comment by manas-yu** [comment]
@Rd4dev PTAL, I have implemented the requested changes and ensured that the jobs now explicitly fail when the status is `action_required`. You can review the updated workflow file [comment_coverage_report.yml](https://github.com/nifux12/oppia-android/blob/develop/.github/workflows/comment_coverage_report.yml) and check the [test pull request](https://github.com/nifux12/oppia-android/pull/7).

Let me know if you need any further changes or clarifications.

**Comment by Rd4dev** [comment]
@manas-yu thanks for promptly addressing the reviews. The changes look good to me!

@adhiamboperes, I've took an initial pass on the PR. Could you please review it and confirm if everything looks good?

**Review by adhiamboperes — APPROVED** [review]
Thanks @manas-yu, @Rd4dev! This LGTM.

**Pull request merged**
